### ISSUE TITLE
preactivation resnet

### ISSUE THREAD

**Issue by bkj** [kuangliu/pytorch-cifar#9]
Hi --

I was looking at the implementation of the preactivation ResNet and had a question -- in the paper, they show the preactivation block below (on the right):
<img width="176" alt="screen shot 2017-07-22 at 3 57 48 pm" src="https://user-images.githubusercontent.com/6086781/28494249-97e81166-6ef6-11e7-88b8-fa4aa184bc0b.png">

which AFAICT should look like:
```
    def forward(self, x):
        shortcut = self.shortcut(x)
        out = self.bn1(x)
        out = F.relu(out)
        out = self.conv1(out)
        
        out = self.bn2(out)
        out = F.relu(out)
        out = self.conv2(out)
        
        out += shortcut
        return out
```
which is slightly different from your implementation:
```
    def forward(self, x):
        out = self.bn1(x)
        out = F.relu(out)
        shortcut = self.shortcut(out)
        out = self.conv1(out)
        
        out = self.bn2(out)
        out = F.relu(out)
        out = self.conv2(out)
        
        out += shortcut
        return out
```
In the picture, I think your implementation would look like shifting the first `bn` and `relu` into the grey band, then splitting into two branches.  

Any thoughts?  Am I misinterpreting, or is there a reason for the difference?

**Comment by kuangliu** [comment]
I think you are right. I kinda forget why I do this...
I'll check the paper later. Thanks.

**Comment by kuangliu** [comment]
The reason why I do this is because I wanna reuse the `x-bn-relu` output.
Notice the shortcut is just `conv`, it should be `bn-relu-conv`, I do `bn-relu` in advance, so the output could be used by both two branches.

**Comment by bkj** [comment]
That makes sense for when `stride != 1 or in_planes != self.expansion*planes`) and there's a `conv` in the shortcut, but when `stride == 1 and in_planes == self.expansion*planes`, then I think the shortcut is just supposed to be the identity.  (I think) that's how it's implemented in the official code:

https://github.com/KaimingHe/resnet-1k-layers/blob/master/resnet-pre-act.lua

(I haven't used Torch before and am partially guessing at the syntax..)

Anyway, it's probably a minor thing and don't want to waste time -- I'll run both versions of the model in a little bit and report back if it makes a difference.

**Comment by kuangliu** [comment]
I checked the official implementation, I think you're right.
Although my implementation works fine, but I think I should respect the official paper implementation.
I'll make a commit changing it later, thanks.

**Comment by kuangliu** [comment]
I'm closing this for now. Please feel free to reopen if you get any interesting result.

**Comment by bkj** [comment]
@kuangliu 

I ran your ResNet18 code as-is once and got a max test accuracy of `94.78`. Then I changed the shortcut as I described above, and got a max test accuracy of `95.03`.   

(Not sure what the variation between runs of each model is, so don't know whether this is a significant difference, but figured I'd report what I'd found so far)

**Comment by kuangliu** [comment]
Cool, `95.03` is very good result. Thanks for the report.
